vm/devices/serial/serial_debugcon/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+                    tracing::info!("debugcon output broken pipe");
+                }
+                Err(err) => {
+                    tracelimit::error_ratelimited!(
+                        len = buf.len(),
+                        error = &err as &dyn std::error::Error,
+                        "debugcon write failed, dropping data"
+                    );
+                    self.tx_buffer.drain(..buf.len());
+                }
+            }
+        }
+        // Wait for more bytes to write.
+        self.tx_waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+impl ChangeDeviceState for SerialDebugcon {
+    fn start(&mut self) {}
+
+    async fn stop(&mut self) {}
+
+    async fn reset(&mut self) {
+        self.tx_buffer.clear();
+    }
+}
+
+impl ChipsetDevice for SerialDebugcon {
+    fn supports_pio(&mut self) -> Option<&mut dyn PortIoIntercept> {
+        Some(self)
+    }
+
+    fn supports_poll_device(&mut self) -> Option<&mut dyn PollDevice> {
+        Some(self)
+    }
+}
+
+impl PollDevice for SerialDebugcon {
+    fn poll_device(&mut self, cx: &mut Context<'_>) {
+        let _ = self.poll_tx(cx);
+        self.sync();
+    }
+}
+
+impl PortIoIntercept for SerialDebugcon {
+    fn io_read(&mut self, io_port: u16, data: &mut [u8]) -> IoResult {
+        if io_port != self.io_port {
+            return IoResult::Err(IoError::InvalidRegister);
+        }
+
+        if data.len() != 1 {
+            return IoResult::Err(IoError::InvalidAccessSize);
+        }
+
+        // this is a magic constant which the guest can use to detect the
+        // presence of the debugcon device. Its value is fixed, and matches the
+        // value set by the original debugcon implementation in bochs.
+        data[0] = 0xe9;
+
+        IoResult::Ok
+    }
+
+    fn io_write(&mut self, io_port: u16, data: &[u8]) -> IoResult {
+        if io_port != self.io_port {
+            return IoResult::Err(IoError::InvalidRegister);
+        }
+
+        if data.len() != 1 {
+            return IoResult::Err(IoError::InvalidAccessSize);
+        }
+
+        self.tx_buffer.push_back(data[0]);
+
+        // HACK: work around the fact that in openvmm, the console is in raw mode.
+        //
+        // FUTURE: this should be configurable, in case folks need 1:1 faithful
+        // debugcon output.
+        if data[0] == b'\n' {
+            self.tx_buffer.push_back(b'\r')
+        }
+
+        self.sync();
+
+        IoResult::Ok
+    }
+
+    fn get_static_regions(&mut self) -> &[(&str, RangeInclusive<u16>)] {
+        std::slice::from_ref(&self.io_region)
+    }
+}
+
+mod save_restore {
+    use crate::SerialDebugcon;
+    use vmcore::save_restore::RestoreError;
+    use vmcore::save_restore::SaveError;
+    use vmcore::save_restore::SaveRestore;
+    use vmcore::save_restore::SavedStateNotSupported;
+
+    impl SaveRestore for SerialDebugcon {
+        // This device should be constructed with `omit_saved_state`.
```

Comment:
I think you want `NoSavedState`, since the presence of this device is visible to the guest (via the 0xe9 read), and so the saved state should reflect that the device needs to be there after a restore.

Reply:
This ties into the comment I made above wrt. this being a debug device, that I wouldn't expect to ever be active in a production scenario. That said - your logic still applies, so I'm open to making this `NoSavedState` instead.

Reply:
It's also a nice sample device, so we probably want to avoid behavior we don't want to see crop up in other devices.

Reply:
Fair point. I'll switch it over.

Reply:
done